File: solutions/csharp/run-length-encoding/1/RunLengthEncoding.cs
using System.Linq;
using System.Text.RegularExpressions;

public static partial class RunLengthEncoding
{
    private static readonly Regex _encodeRegex = EncodeRegex();
    private static readonly Regex _decodeRegex = DecodeRegex();

    public static string Encode(string input) => string.Join(string.Empty, _encodeRegex.Matches(input)
                                                                                      .Select(m => $"{(m.Length > 1 ? m.Length.ToString() : string.Empty)}{m.Value[0]}"));

    public static string Decode(string input) => string.Join(string.Empty, _decodeRegex.Matches(input)
                                                                                       .Select(m => m.Length < 2 ? m.Value : new(m.Value[^1], int.Parse(m.Value[..^1]))));

    [GeneratedRegex("(.)\\1*")]
    private static partial Regex EncodeRegex();

    [GeneratedRegex(@"(\d*.)")]
    private static partial Regex DecodeRegex();
}

File: solutions/julia/run-length-encoding/1/run-length-encoding.jl
function encode(s)
    length(s) < 2 && return s

    s *= "."
    ccount, cchar, output = 1, s[begin], ""

    for c in s[2:end]
        if cchar == c
            ccount += 1
        else
            output *= "$(ccount > 1 ? ccount : "")$cchar"
            ccount, cchar = 1, c
        end
    end

    return output
end

function decode(s::AbstractString)
    output = cn = ""

    for c in s
        if !isnumeric(c)
            cn = isempty(cn) ? "1" : cn
            output *= c ^ parse(Int, cn)
            cn = ""
        else
            cn *= c
        end
    end

    return output
end